[WebXR] Implement explicit resolve of WebXR MSAA Rendertarget

djg opened the pull request:
#### 606967dd83de240185d0a057c06725572478321d
<pre>
[WebXR] Implement explicit resolve of WebXR MSAA Rendertarget
<a href="https://bugs.webkit.org/show_bug.cgi?id=274137">https://bugs.webkit.org/show_bug.cgi?id=274137</a>
<a href="https://rdar.apple.com/128041507">rdar://128041507</a>

Reviewed by Mike Wyrzykowski.

GL requires a glBlitFramebuffer to resolve MSAA render targets into
single-sample render targets. Internally, ANGLE&apos;s Metal backend has an implicit
resolve target that it allocates, which becomes the source for the
glBlitFramebuffer operation. GPU profiling data captured from device shows that
a large amount of time and memory bandwidth is being used to store both MSAA and
single-sample render targets, as well as, the blit operation. We would like to
avoid this and use Metal&apos;s render pass resolve target and discard the MSAA data
at the end of rendering.

In lieu of ANGLE&apos;s Metal backend being able to optimize this GL API usage
pattern to use render pass resolve target, this change makes use of a custom
extension to explicitly set the renderbuffer to use as resolve target in
the Metal renderer instead of the implicit render target, with:

    glFramebufferResolveRenderbufferWEBKIT(...)

Using this API removes the need to use glBlitFramebuffer variants to resolve
multiple samples to a single sample, and in combination with
glDiscardFramebufferEXT/glInvalidateFramebuffer, can avoid the cost of storing
MSAA buffers at the end of rendering.

This is not an long term solution and not intended for upstreaming to the ANGLE
project, as we&apos;re not interested in supporting all the combinations a general
API would require. The long term solution is to teach ANGLE&apos;s Metal backend to
optimize away the glBlitFramebuffer resolve step, just as ANGLE&apos;s Vulkan backend
is able to do.

* Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj:
* Source/ThirdParty/ANGLE/Configurations/BaseTarget.xcconfig:
* Source/ThirdParty/ANGLE/doc/ExtensionSupport.md:
* Source/ThirdParty/ANGLE/include/GLES2/gl2ext_angle.h:
* Source/ThirdParty/ANGLE/scripts/code_generation_hashes/GL_EGL_entry_points.json:
* Source/ThirdParty/ANGLE/scripts/code_generation_hashes/interpreter_utils.json:
* Source/ThirdParty/ANGLE/scripts/entry_point_packed_gl_enums.json:
* Source/ThirdParty/ANGLE/scripts/gl_angle_ext.xml:
* Source/ThirdParty/ANGLE/scripts/registry_xml.py:
* Source/ThirdParty/ANGLE/src/common/entry_points_enum_autogen.cpp:
(angle::GetEntryPointName):
* Source/ThirdParty/ANGLE/src/common/entry_points_enum_autogen.h:
* Source/ThirdParty/ANGLE/src/libANGLE/Context.cpp:
(gl::Context::framebufferResolveRenderbufferWEBKIT):
* Source/ThirdParty/ANGLE/src/libANGLE/Context_gles_ext_autogen.h:
* Source/ThirdParty/ANGLE/src/libANGLE/Framebuffer.cpp:
(gl::FramebufferState::FramebufferState):
(gl::FramebufferState::getColorResolveAttachment const):
(gl::FramebufferState::getDepthResolveAttachment const):
(gl::FramebufferState::getStencilResolveAttachment const):
(gl::Framebuffer::Framebuffer):
(gl::Framebuffer::onDestroy):
(gl::Framebuffer::getColorResolveAttachment const):
(gl::Framebuffer::getDepthResolveAttachment const):
(gl::Framebuffer::getStencilResolveAttachment const):
(gl::Framebuffer::setAttachmentResolve):
(gl::Framebuffer::updateAttachmentResolve):
(gl::Framebuffer::resetAttachmentResolve):
* Source/ThirdParty/ANGLE/src/libANGLE/Framebuffer.h:
* Source/ThirdParty/ANGLE/src/libANGLE/capture/capture_gles_ext_autogen.cpp:
(gl::CaptureFramebufferResolveRenderbufferWEBKIT):
* Source/ThirdParty/ANGLE/src/libANGLE/capture/capture_gles_ext_autogen.h:
* Source/ThirdParty/ANGLE/src/libANGLE/gles_extensions_autogen.cpp:
(gl::GetExtensionInfoMap):
* Source/ThirdParty/ANGLE/src/libANGLE/gles_extensions_autogen.h:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/BUILD.gn:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/ContextMtl.mm:
(rx::ContextMtl::bindMetalRasterizationRateMap):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/DisplayMtl.mm:
(rx::DisplayMtl::initializeExtensions const):
(rx::DisplayMtl::initializeFeatures):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/FrameBufferMtl.h:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/FrameBufferMtl.mm:
(rx::FramebufferMtl::FramebufferMtl):
(rx::FramebufferMtl::reset):
(rx::FramebufferMtl::setLoadStoreActionOnRenderPassFirstStart):
(rx::FramebufferMtl::updateColorRenderTarget):
(rx::FramebufferMtl::updateDepthRenderTarget):
(rx::FramebufferMtl::updateStencilRenderTarget):
(rx::FramebufferMtl::prepareRenderPass):
(rx::FramebufferMtl::clearWithLoadOpRenderPassNotStarted):
(rx::FramebufferMtl::invalidateImpl):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/RenderTargetMtl.h:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/RenderTargetMtl.mm:
(rx::RenderTargetMtl::toRenderPassAttachmentDesc const):
(rx::RenderTargetMtl::toRenderPassResolveAttachmentDesc const):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_command_buffer.mm:
(rx::mtl::RenderCommandEncoder::initAttachmentWriteDependencyAndScissorRect):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_state_cache.h:
(rx::mtl::RenderPassAttachmentDesc::hasResolveTexture const):
(rx::mtl::RenderPassAttachmentDesc::hasImplicitMSTexture const): Deleted.
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_state_cache.mm:
(rx::mtl::RenderPassAttachmentDesc::reset):
(rx::mtl::RenderPassAttachmentDesc::equalIgnoreLoadStoreOptions const):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_utils.mm:
* Source/ThirdParty/ANGLE/src/libANGLE/validationESEXT.cpp:
(gl::ValidateFramebufferResolveRenderbufferWEBKIT):
* Source/ThirdParty/ANGLE/src/libANGLE/validationESEXT_autogen.h:
* Source/ThirdParty/ANGLE/src/libGLESv2/entry_points_gles_ext_autogen.cpp:
* Source/ThirdParty/ANGLE/src/libGLESv2/entry_points_gles_ext_autogen.h:
* Source/ThirdParty/ANGLE/src/libGLESv2/libGLESv2_autogen.cpp:
* Source/ThirdParty/ANGLE/src/libGLESv2/libGLESv2_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/libGLESv2_no_capture_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/libGLESv2_vulkan_secondaries_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/libGLESv2_with_capture_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/opengl32_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/opengl32_with_wgl_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/proc_table_egl_autogen.cpp:
* Source/ThirdParty/ANGLE/src/libGLESv2/proc_table_glx_autogen.cpp:
* Source/ThirdParty/ANGLE/src/libGLESv2/proc_table_wgl_autogen.cpp:
* Source/ThirdParty/ANGLE/src/tests/angle_end2end_tests.gni:
* Source/ThirdParty/ANGLE/util/capture/frame_capture_replay_autogen.cpp:
(angle::ReplayTraceFunctionCall):
* Source/ThirdParty/ANGLE/util/capture/trace_gles_loader_autogen.cpp:
* Source/ThirdParty/ANGLE/util/capture/trace_gles_loader_autogen.h:
* Source/ThirdParty/ANGLE/util/capture/trace_interpreter_autogen.cpp:
(angle::ParseCallCapture):
* Source/ThirdParty/ANGLE/util/gles_loader_autogen.cpp:
* Source/ThirdParty/ANGLE/util/gles_loader_autogen.h:
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
* Source/WebCore/Modules/webxr/WebXROpaqueFramebufferCocoa.cpp:
(WebCore::WebXROpaqueFramebuffer::resolveMSAAFramebuffer):
(WebCore::WebXROpaqueFramebuffer::setupFramebuffer):
(WebCore::WebXROpaqueFramebuffer::bindResolveAttachments):
* Source/WebCore/platform/graphics/GraphicsContextGL.cpp:
(WebCore::GraphicsContextGL::framebufferResolveRenderbuffer):
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLCocoa::framebufferResolveRenderbuffer):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(framebufferResolveRenderbuffer):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGLProxy::framebufferResolveRenderbuffer):

Canonical link: <a href="https://commits.webkit.org/280523@main">https://commits.webkit.org/280523@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2c538ae2769a99f5bf45acd0932847997699314a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56879 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36207 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9353 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60498 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7322 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59007 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43830 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7512 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46070 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5137 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58909 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34009 "Found 60 new test failures: accessibility/accessibility-node-reparent.html accessibility/ancestor-computation.html accessibility/announcement-notification.html accessibility/aria-braillelabel.html accessibility/aria-brailleroledescription.html accessibility/aria-checked-mixed-value.html accessibility/aria-current-state-changed-notification.html accessibility/aria-current.html accessibility/aria-describedby-on-input.html accessibility/aria-description.html ... (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49089 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26928 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30789 "Found 60 new test failures: imported/w3c/web-platform-tests/FileAPI/Blob-methods-from-detached-frame.html imported/w3c/web-platform-tests/FileAPI/FileReader/Progress_event_bubbles_cancelable.html imported/w3c/web-platform-tests/FileAPI/FileReader/workers.html imported/w3c/web-platform-tests/FileAPI/FileReaderSync.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-array-buffer.any.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-array-buffer.any.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-constructor-dom.window.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-constructor-endings.html ... (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6420 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6326 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/52744 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6691 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62180 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/792 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6795 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53325 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/795 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49144 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53361 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/12569 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/671 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8469 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32036 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33121 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34206 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32867 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/ThirdParty/ANGLE/src/libANGLE/FramebufferAttachment.h:
```diff
@@ -190,17 +198,17 @@ class FramebufferAttachment final
-    GLenum mType;
-    Target mTarget;
-    FramebufferAttachmentObject *mResource;
+    GLenum mBinding;
```
It would seem to me that Target is a pair (binding, textureIndex) before and after your change.

It would seem that your problem is that you observed "binding is the same, type differs".
Maybe you could have less changes by preserving the Target type as is, and have
`GLenum mType[kSlotCount];`


and even less changes you'd have:

```
...
GLenum mResolveType =GL_NONE;
Target mResolveTarget;
FramebufferAttachmentObject *mResolveResource = nullptr;
```

Then for the few functions that need the slot, add a function that returns `return slot == kMainSlot? mResource : mResolveResource`

kkinnunen-apple commented on Source/ThirdParty/ANGLE/src/libANGLE/FramebufferAttachment.h:
```diff
@@ -190,17 +198,17 @@ class FramebufferAttachment final
         Target(const Target &other);
```
Above, you forgot to rename `binding`..

djg commented on Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj:
```diff
@@ -3,7 +3,7 @@
 	classes = {
 	};
-	objectVersion = 55;
+	objectVersion = 54;
```
Why does Xcode do this?

djg commented on Source/ThirdParty/ANGLE/src/libANGLE/Framebuffer.h:
```diff
@@ -162,6 +169,13 @@ class FramebufferState final : angle::NonCopyable
     FramebufferAttachment mDepthAttachment;
     FramebufferAttachment mStencilAttachment;
 
+#if ANGLE_WEBKIT_EXPLICIT_RESOLVE_TARGET_ENABLED
+    // GL_WEBKIT_explicit_resolve_target;
+    DrawBuffersVector<FramebufferAttachment> mColorResolveAttachments;
+    FramebufferAttachment mDepthResolveAttachment;
+    FramebufferAttachment mStencilResolveAttachment;
+#endif
```
To address Kimmo's comment, instead of adding an extra slot to `FramebufferAttachment`, add separate storage for the resolve target attachment

djg commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_state_cache.h:
```diff
@@ -319,15 +319,19 @@ struct RenderPassAttachmentDesc
     TextureRef texture;
-    // Implicit multisample texture that will be rendered into and discarded at the end of
-    // a render pass. Its result will be resolved into normal texture above.
-    TextureRef implicitMSTexture;
```
It seemed silly to me to have the target for drawing be either texture or implicitMSTexture if it is set.

djg commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_state_cache.h:
```diff
@@ -319,15 +319,19 @@ struct RenderPassAttachmentDesc
     TextureRef texture;
-    // Implicit multisample texture that will be rendered into and discarded at the end of
-    // a render pass. Its result will be resolved into normal texture above.
-    TextureRef implicitMSTexture;
+    TextureRef resolveTexture;
```
With this change, `texture` is always the destination for rendering and `resolveTexture` is always the destination for the automatic resolve step.

I was going to change `texture` to `drawingTexture` but that added a lot of `renaming of member` to the patch, which I decided add too much noise.

djg commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_state_cache.mm:
```diff
@@ -699,19 +684,21 @@ MTLColorWriteMask AdjustColorWriteMaskForSharedExponent(MTLColorWriteMask mask)
 {
     texture.reset();
-    implicitMSTexture.reset();
-    level              = mtl::kZeroNativeMipLevel;
-    sliceOrDepth       = 0;
```
This state was either for `implicitMSTexture` or `texture`.

djg commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_state_cache.mm:
```diff
@@ -699,19 +684,21 @@ MTLColorWriteMask AdjustColorWriteMaskForSharedExponent(MTLColorWriteMask mask)
+    level               = mtl::kZeroNativeMipLevel;
+    sliceOrDepth        = 0;
+    resolveLevel        = mtl::kZeroNativeMipLevel;
+    resolveSliceOrDepth = 0;
```
Now the state for each target is explicitly for the one it's intended to be used with.

djg commented on Source/ThirdParty/ANGLE/src/libANGLE/validationESEXT.cpp:
```diff
@@ -4461,6 +4461,30 @@ bool ValidateLogicOpANGLE(const PrivateState &state,
+    }
+
+    // TODO(djg): Check compatibility with default
+    return ValidateFramebufferRenderbufferBase(context, entryPoint, target, attachment,
```
In an upstreamed extension this would require a lot more validation, but we're controlling the use of the API and not exposing it publicly.

kkinnunen-apple replied:
But an attacker could call it with arbitrary arguments..

djg replied:
I've added validation of compatibility of the attachments to glCheckFramebufferStatus, since at that point we have both attachments to compare with each other. This validation is the equivalent of the validation performed by `glFramebufferRenderbuffer(...)`

kkinnunen-apple reviewed (commented):
A lot nicer implementation.

djg committed:
[WebXR] Implement explicit resolve of WebXR MSAA Rendertarget
https://bugs.webkit.org/show_bug.cgi?id=274137
rdar://128041507

Reviewed by Mike Wyrzykowski.

GL requires a glBlitFramebuffer to resolve MSAA render targets into
single-sample render targets. Internally, ANGLE's Metal backend has an implicit
resolve target that it allocates, which becomes the source for the
glBlitFramebuffer operation. GPU profiling data captured from device shows that
a large amount of time and memory bandwidth is being used to store both MSAA and
single-sample render targets, as well as, the blit operation. We would like to
avoid this and use Metal's render pass resolve target and discard the MSAA data
at the end of rendering.

In lieu of ANGLE's Metal backend being able to optimize this GL API usage
pattern to use render pass resolve target, this change makes use of a custom
extension to explicitly set the renderbuffer to use as resolve target in
the Metal renderer instead of the implicit render target, with:

    glFramebufferResolveRenderbufferWEBKIT(...)

Using this API removes the need to use glBlitFramebuffer variants to resolve
multiple samples to a single sample, and in combination with
glDiscardFramebufferEXT/glInvalidateFramebuffer, can avoid the cost of storing
MSAA buffers at the end of rendering.

This is not an long term solution and not intended for upstreaming to the ANGLE
project, as we're not interested in supporting all the combinations a general
API would require. The long term solution is to teach ANGLE's Metal backend to
optimize away the glBlitFramebuffer resolve step, just as ANGLE's Vulkan backend
is able to do.

* Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj:
* Source/ThirdParty/ANGLE/Configurations/BaseTarget.xcconfig:
* Source/ThirdParty/ANGLE/doc/ExtensionSupport.md:
* Source/ThirdParty/ANGLE/include/GLES2/gl2ext_angle.h:
* Source/ThirdParty/ANGLE/scripts/code_generation_hashes/GL_EGL_entry_points.json:
* Source/ThirdParty/ANGLE/scripts/code_generation_hashes/interpreter_utils.json:
* Source/ThirdParty/ANGLE/scripts/entry_point_packed_gl_enums.json:
* Source/ThirdParty/ANGLE/scripts/gl_angle_ext.xml:
* Source/ThirdParty/ANGLE/scripts/registry_xml.py:
* Source/ThirdParty/ANGLE/src/common/entry_points_enum_autogen.cpp:
(angle::GetEntryPointName):
* Source/ThirdParty/ANGLE/src/common/entry_points_enum_autogen.h:
* Source/ThirdParty/ANGLE/src/libANGLE/Context.cpp:
(gl::Context::framebufferResolveRenderbufferWEBKIT):
* Source/ThirdParty/ANGLE/src/libANGLE/Context_gles_ext_autogen.h:
* Source/ThirdParty/ANGLE/src/libANGLE/Framebuffer.cpp:
(gl::FramebufferState::FramebufferState):
(gl::FramebufferState::getColorResolveAttachment const):
(gl::FramebufferState::getDepthResolveAttachment const):
(gl::FramebufferState::getStencilResolveAttachment const):
(gl::Framebuffer::Framebuffer):
(gl::Framebuffer::onDestroy):
(gl::Framebuffer::getColorResolveAttachment const):
(gl::Framebuffer::getDepthResolveAttachment const):
(gl::Framebuffer::getStencilResolveAttachment const):
(gl::Framebuffer::setAttachmentResolve):
(gl::Framebuffer::updateAttachmentResolve):
(gl::Framebuffer::resetAttachmentResolve):
* Source/ThirdParty/ANGLE/src/libANGLE/Framebuffer.h:
* Source/ThirdParty/ANGLE/src/libANGLE/capture/capture_gles_ext_autogen.cpp:
(gl::CaptureFramebufferResolveRenderbufferWEBKIT):
* Source/ThirdParty/ANGLE/src/libANGLE/capture/capture_gles_ext_autogen.h:
* Source/ThirdParty/ANGLE/src/libANGLE/gles_extensions_autogen.cpp:
(gl::GetExtensionInfoMap):
* Source/ThirdParty/ANGLE/src/libANGLE/gles_extensions_autogen.h:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/BUILD.gn:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/ContextMtl.mm:
(rx::ContextMtl::bindMetalRasterizationRateMap):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/DisplayMtl.mm:
(rx::DisplayMtl::initializeExtensions const):
(rx::DisplayMtl::initializeFeatures):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/FrameBufferMtl.h:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/FrameBufferMtl.mm:
(rx::FramebufferMtl::FramebufferMtl):
(rx::FramebufferMtl::reset):
(rx::FramebufferMtl::setLoadStoreActionOnRenderPassFirstStart):
(rx::FramebufferMtl::updateColorRenderTarget):
(rx::FramebufferMtl::updateDepthRenderTarget):
(rx::FramebufferMtl::updateStencilRenderTarget):
(rx::FramebufferMtl::prepareRenderPass):
(rx::FramebufferMtl::clearWithLoadOpRenderPassNotStarted):
(rx::FramebufferMtl::invalidateImpl):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/RenderTargetMtl.h:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/RenderTargetMtl.mm:
(rx::RenderTargetMtl::toRenderPassAttachmentDesc const):
(rx::RenderTargetMtl::toRenderPassResolveAttachmentDesc const):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_command_buffer.mm:
(rx::mtl::RenderCommandEncoder::initAttachmentWriteDependencyAndScissorRect):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_state_cache.h:
(rx::mtl::RenderPassAttachmentDesc::hasResolveTexture const):
(rx::mtl::RenderPassAttachmentDesc::hasImplicitMSTexture const): Deleted.
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_state_cache.mm:
(rx::mtl::RenderPassAttachmentDesc::reset):
(rx::mtl::RenderPassAttachmentDesc::equalIgnoreLoadStoreOptions const):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_utils.mm:
* Source/ThirdParty/ANGLE/src/libANGLE/validationESEXT.cpp:
(gl::ValidateFramebufferResolveRenderbufferWEBKIT):
* Source/ThirdParty/ANGLE/src/libANGLE/validationESEXT_autogen.h:
* Source/ThirdParty/ANGLE/src/libGLESv2/entry_points_gles_ext_autogen.cpp:
* Source/ThirdParty/ANGLE/src/libGLESv2/entry_points_gles_ext_autogen.h:
* Source/ThirdParty/ANGLE/src/libGLESv2/libGLESv2_autogen.cpp:
* Source/ThirdParty/ANGLE/src/libGLESv2/libGLESv2_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/libGLESv2_no_capture_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/libGLESv2_vulkan_secondaries_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/libGLESv2_with_capture_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/opengl32_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/opengl32_with_wgl_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/proc_table_egl_autogen.cpp:
* Source/ThirdParty/ANGLE/src/libGLESv2/proc_table_glx_autogen.cpp:
* Source/ThirdParty/ANGLE/src/libGLESv2/proc_table_wgl_autogen.cpp:
* Source/ThirdParty/ANGLE/src/tests/angle_end2end_tests.gni:
* Source/ThirdParty/ANGLE/util/capture/frame_capture_replay_autogen.cpp:
(angle::ReplayTraceFunctionCall):
* Source/ThirdParty/ANGLE/util/capture/trace_gles_loader_autogen.cpp:
* Source/ThirdParty/ANGLE/util/capture/trace_gles_loader_autogen.h:
* Source/ThirdParty/ANGLE/util/capture/trace_interpreter_autogen.cpp:
(angle::ParseCallCapture):
* Source/ThirdParty/ANGLE/util/gles_loader_autogen.cpp:
* Source/ThirdParty/ANGLE/util/gles_loader_autogen.h:
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
* Source/WebCore/Modules/webxr/WebXROpaqueFramebufferCocoa.cpp:
(WebCore::WebXROpaqueFramebuffer::resolveMSAAFramebuffer):
(WebCore::WebXROpaqueFramebuffer::setupFramebuffer):
(WebCore::WebXROpaqueFramebuffer::bindResolveAttachments):
* Source/WebCore/platform/graphics/GraphicsContextGL.cpp:
(WebCore::GraphicsContextGL::framebufferResolveRenderbuffer):
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLCocoa::framebufferResolveRenderbuffer):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(framebufferResolveRenderbuffer):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGLProxy::framebufferResolveRenderbuffer):

Canonical link: https://commits.webkit.org/280523@main
